### PR TITLE
Add useKeyOrderedProcessing option to DefaultReactivePulsarMessageListenerContainer

### PR DESCRIPTION
**Pull request by cbornet**

### CODE REVIEW

File: spring-pulsar/src/main/java/org/springframework/pulsar/listener/reactive/DefaultReactivePulsarMessageListenerContainer.java [hunk 1]
```diff
@@ -179,8 +180,13 @@ private ReactiveMessagePipeline startPipeline(ReactivePulsarContainerProperties<
 					.messageHandler(((ReactivePulsarOneByOneMessageHandler<T>) messageHandler)::received)
 					.handlingTimeout(containerProperties.getHandlingTimeout());
 			if (containerProperties.getConcurrency() > 0) {
-				pipeline = messagePipelineBuilder.concurrent().concurrency(containerProperties.getConcurrency())
-						.maxInflight(containerProperties.getMaxInFlight()).build();
+				ConcurrentOneByOneMessagePipelineBuilder<T> concurrentPipelineBuilder = messagePipelineBuilder
+						.concurrent().concurrency(containerProperties.getConcurrency())
+						.maxInflight(containerProperties.getMaxInFlight());
+				if (containerProperties.isUseKeyOrderedProcessing()) {
+					concurrentPipelineBuilder.useKeyOrderedProcessing();
```
Comment (onobc) [inline spring-pulsar/src/main/java/org/springframework/pulsar/listener/reactive/DefaultReactivePulsarMessageListenerContainer.java:187]: **[NOT-DIRECTLY-RELATED-WISHLIST]** The fluent builder syntax is great.... until you have to do something conditionally. I want to explore a builder of the form `field(T)` and `fieldIf(Supplier<T>, Condition<T>)` sometime. 

Ok, please resolve this comment once digested. Just sharing thoughts.

File: spring-pulsar/src/main/java/org/springframework/pulsar/listener/reactive/ReactivePulsarContainerProperties.java [hunk 2]
```diff
@@ -49,6 +49,8 @@
 
 	private int maxInFlight = 0;
 
+	private boolean useKeyOrderedProcessing = false;
```
Comment (onobc) [inline spring-pulsar/src/main/java/org/springframework/pulsar/listener/reactive/ReactivePulsarContainerProperties.java:52]: This reminds me that there is probably nothing stopping us from turning this into a simple `record`.

File: spring-pulsar/src/test/java/org/springframework/pulsar/listener/reactive/DefaultReactivePulsarMessageListenerContainerTests.java [hunk 3]
```diff
@@ -158,7 +160,8 @@ void containerProperties() throws Exception {
 
 		assertThat(container).extracting("pipeline", InstanceOfAssertFactories.type(ReactiveMessagePipeline.class))
 				.hasFieldOrPropertyWithValue("concurrency", 5).hasFieldOrPropertyWithValue("maxInflight", 6)
-				.hasFieldOrPropertyWithValue("handlingTimeout", Duration.ofMillis(7));
+				.hasFieldOrPropertyWithValue("handlingTimeout", Duration.ofMillis(7)).extracting("groupingFunction")
```
Comment (onobc) [inline spring-pulsar/src/test/java/org/springframework/pulsar/listener/reactive/DefaultReactivePulsarMessageListenerContainerTests.java:163]: Later we should consider breaking this test into smaller pieces as it is not directly clear that "setUseKeyOrderedProcessing" equates to "groupFunction == DMGF".

**Review by onobc — APPROVED** [review]

File: spring-pulsar/src/main/java/org/springframework/pulsar/listener/reactive/DefaultReactivePulsarMessageListenerContainer.java (hunk 1 above)
Comment (cbornet) [reply spring-pulsar/src/main/java/org/springframework/pulsar/listener/reactive/DefaultReactivePulsarMessageListenerContainer.java:187]: cc @lhotari

**Pull request merged**